src/compiler/checker.ts

```diff
@@ -5345,72 +5333,61 @@ namespace ts {
                 case SyntaxKind.NullKeyword:
                 case SyntaxKind.NeverKeyword:
                 case SyntaxKind.LiteralType:
-                    return true;
+                    return false;
                 case SyntaxKind.ArrayType:
-                    return isThislessType((<ArrayTypeNode>node).elementType);
+                    return typeMayReferenceThis((<ArrayTypeNode>node).elementType);
                 case SyntaxKind.TypeReference:
-                    return !(node as TypeReferenceNode).typeArguments || (node as TypeReferenceNode).typeArguments.every(isThislessType);
+                    return some((node as TypeReferenceNode).typeArguments, typeMayReferenceThis);
             }
-            return false;
-        }
-
-        /** A type parameter is thisless if its contraint is thisless, or if it has no constraint. */
-        function isThislessTypeParameter(node: TypeParameterDeclaration) {
-            return !node.constraint || isThislessType(node.constraint);
+            return true;

```

Comment:
should probably reference #20034 here too, because there are plenty of type nodes that could be inspected for `this` more closely.